### PR TITLE
[AIT-1146] Add first spec-derived UTS integration and proxy tests via rewritten uts-to-swift skill

### PR DESCRIPTION
**Pull request by sacOO7**
- Fixes https://ably.atlassian.net/browse/AIT-1146

## 🎯 What this PR does

Rewrites the **`uts-to-swift` skill** (v2.0.0) to mirror ably-java's [`uts-to-kotlin`](https://github.com/ably/ably-java/tree/main/.claude/skills/uts-to-kotlin) step for step. The skill previously translated **one spec file at a time** with hand-maintained path tables and an eyeball-only review; it now drives the **whole module-directory workflow** the kotlin skill established, backed by two deterministic scripts:

```
/uts-to-swift <path-to-uts-module-directory>     # e.g. <cloned-ably-specification-repo-path>/uts/realtime
```

Built on top of the `Test/UTS` integration/proxy infrastructure introduced by #2223 (hence the base branch).

The PR also ships the skill's **first real output**: three spec-derived integration tests that replace the interim smoke tests, with all env-var gating dropped and the UTS docs brought to block-level parity with ably-java — see the [spec-derived tests section](#-first-spec-derived-integration-tests-58fa0f5e) below.

## 🧭 The new two-phase flow

**Phase 1 — Selection** (scripted, byte-for-byte deterministic):

1. `scripts/resolve_uts.py` validates the module directory, reads `uts-package-mapping.json`, and emits per-tier target directories plus every candidate spec with its derived Swift class name (`connection_recovery_test.md` → `ConnectionRecoveryTests`).
2. The user then confirms/creates the module mapping, picks a **tier** (unit / integration / proxy), picks **which specs**, and chooses **translate-only** vs **translate-and-evaluate**.

**Phase 2 — Per-spec translation** (Steps 1–7): read spec (+ per-module translation notes) → resolver-provided output path → read the tier's infra → generate → compile → run *(evaluate mode only)* → audit-driven review.

## 📦 What's new alongside SKILL.md

| File | Purpose |
|---|---|
| `scripts/resolve_uts.py` | Module validation + tier/spec/class-name resolution; `--create` maintains the mapping. Minimal-diff port of kotlin's (only deltas: `Tests` suffix, no Kotlin package concept, cocoa wording). |
| `scripts/audit_translation.py` | Step 7 review aid: Test-ID coverage (`missingInSwift` / `orphanInSwift` from `// UTS:` tags) + a per-test spec-line ledger with an **assertion-shortfall tripwire**. |
| `uts-package-mapping.json` | Spec module → `Test/UTS` tier-directory mapping (`realtime`, `rest`, `objects`). |
| `references/objects-mapping.md` | Intentionally-empty placeholder for the future ably-js → ably-cocoa LiveObjects type map (authored when `objects` translation starts). |

## ✍️ SKILL.md highlights

- **Structure matches kotlin heading-for-heading** — the only structural deltas are Swift necessities, each deliberate: no package concept, scoped-resource base cases instead of `@BeforeAll`/`finally` teardown, and JSON forced explicitly on the proxy tier.
- **All cocoa-specific translation content retained**: the `import Ably.Private` internals-access ladder, `Captured<T>`/Swift-6-Sendable rules, paired pseudocode→Swift mock examples, and per-tier file templates (unit / direct-sandbox / proxy).
- **Evaluation model upgraded**: every run ends in one of three legal terminal states — spec-correct pass, documented SDK deviation (env-gated/adapted), or **UTS spec error failing fast** (`Issue.record`) so the spec gets fixed at source. The accommodate-both pattern stays banned.
- **Required reading is fetched, not remembered**: `writing-derived-tests.md` up front, and `uts/docs/proxy.md` before any proxy-tier translation (with a local-checkout fallback while the upstream `docs/` move is unmerged).
- Documents two known gaps honestly rather than papering over them: the WS mock lacks `respondWithTimeout`/`respondWithDNSError` (extend-or-record instruction), and fixture-driven specs (e.g. `msgpack_interop.md`) legitimately yield `specCount: 0` + an expected orphan tag in the audit.

## 🔍 One deliberate improvement over the kotlin original

`audit_translation.py` **skips comment-only lines** when counting assertions — a commented-out `// #expect(…)` is not an assertion, and counting it (as kotlin's script does with a commented `assertEquals`) would silently mask exactly the dropped-assertion case the tripwire exists to catch. Worth upstreaming to `uts-to-kotlin`.

## ✅ Verification (initial skill commit)

- **Resolver**: exercised against the real spec checkout for all three modules; all five error codes (`NOT_A_UTS_MODULE_PATH`, `DIR_NOT_FOUND`, `NO_TIER_DIRS`, `MAPPING_NOT_FOUND`, `BAD_TARGET_NAME`) verified; `--create` round-trip preserves `notes` and LF endings.
- **Audit**: clean ID coverage on both existing spec/test pairs (`ConnectionRecoveryTests`, `TimeTests` — the one shortfall it flags is `TimeTests`' documented annotated omission); mutation tests confirm missing-test, orphan-tag, and shortfall detection; swept over the **entire spec corpus (146 files) with zero extractor failures**.
- **Skill**: heading skeleton diffs against `uts-to-kotlin` with only the documented deltas; every snippet API verified against the actual `Test/UTS/infra` signatures.
- `swift test --filter UTS` passes; `make lint` clean; the initial commit touches nothing under `Source/` or `Test/` (later commits in this PR do add tests under `Test/UTS/` — see below).

## 🔁 Review follow-ups (`7e383be3`, `455f4ca7`)

All ten review comments across two rounds are addressed in the threads; the code changes:

- **Split assert/await accounting** in the audit (`swiftAssertionCount` vs `swiftAwaitCount`, `assertionShortfall` primary / `awaitShortfall` advisory) — a wait surplus could previously mask deleted assertions (proven on `RTN16f`).
- **Pseudo-only fence extraction** — only ` ```pseudo `-labelled blocks feed the spec ledger (corpus-verified output-neutral today).
- **`duplicateInSwift`** in `idCoverage` + the exit-2 gate — a duplicated `// UTS:` tag previously made a method invisible to the audit (reported, not merged; not a hard failure).
- **`BAD_MAPPING`** fail-fast for a blank/invalid `testRoot` (validated *before* `--create` can persist corruption) and **`expanduser()`** for `~`-paths in the resolver.
- Docstring escape cleanup (Ruff W605) and **username-neutral example paths** (`<cloned-ably-specification-repo-path>/uts/objects`).
- One suggestion deliberately rejected with rationale in-thread: full lexical comment/string stripping in the audit (a mini-lexer in an explicitly advisory review aid).

## 🧪 First spec-derived integration tests (`58fa0f5e`)

The rewritten skill was then used for real (translate-and-evaluate mode) — this PR includes its first output, replacing the interim infra "smoke" tests per their TODOs:

| Test | Spec | Covers |
|---|---|---|
| `integration/standard/realtime/ChannelHistoryTests.swift` | `channel_history_test.md` (RTL10d) | Cross-client history, newest-first, once per protocol variant (JSON + msgpack) |
| `integration/standard/realtime/TokenRequestTests.swift` | `token_request_test.md` (RSA9/RSA9a/RSA9g) | Locally signed `TokenRequest` accepted by the real server, with/without `clientId` |
| `integration/proxy/realtime/AuthReauthTests.swift` | `auth_reauth.md` (RTN22/RTC8a) | Injected server-initiated AUTH → re-auth via `authCallback` + client→server AUTH frame in the proxy log, connection undisrupted. macOS-only |

- **`IntegrationSmokeTest` / `ProxyInfraSmokeTests` removed** — the spec-derived tests cover their ground end-to-end (sandbox provisioning, real clients, proxy chain, log assertions).
- **All env-var gating dropped** (`UTS_INTEGRATION_*`): every tier now runs directly from `swift test` and the **Xcode test runner UI**; tier selection is by `--filter`, mirroring ably-java's `runUtsUnitTests`/`runUtsIntegrationTests` split. `UTS_PROXY_LOCAL_PATH` (config override) and `RUN_DEVIATIONS` (deviation mechanism) remain.
- **`Test/UTS/README.md` brought to block-level parity with ably-java's `uts/README.md`**: §2 tier table with "Example in this repo" paths, §7 walkthrough restructured to numbered sub-sections, §10 rewritten around ungated per-suite runs, §11.4/§11.5 now walk the real `ChannelHistoryTests`/`AuthReauthTests` with code excerpts at the same points as ably-java's §10/§11, §11.3 control-API endpoints + common-actions list, §12 cheat-sheet build-client snippets / Test-ID format / decision-tree entries.
- **SKILL.md**: reference examples now point at the spec-derived suites (matching kotlin's `ChannelHistoryTest` reference), and generated tests are instructed to stay ungated.

**Verification**: all three suites pass ungated against the real sandbox (TokenRequest 2/2, ChannelHistory 2/2 JSON+msgpack, AuthReauth 1/1 incl. proxy binary sync); unit suites 11/11; `swift build --build-tests` and `make lint` clean; zero smoke/env-gate references remain in code or docs.

## Summary by CodeRabbit

* **New Features**
  * Added deterministic tooling to resolve UTS module directories into Swift test targets.
  * Added an automated faithfulness audit that reports missing/orphan/duplicate test identifiers and assertion shortfalls.
  * Added UTS-to-Swift package mappings for realtime, objects, and REST across unit/integration/proxy tiers.

* **Documentation**
  * Substantially rewrote the UTS-to-Swift translation guide, including two-phase workflow, tighter naming/assertion rules, and Swift 6 mocking concurrency guidance.
  * Added an objects mapping reference placeholder.

### CODE REVIEW

File: .claude/skills/uts-to-swift/scripts/audit_translation.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+generated Swift test, so the per-spec review (SKILL.md Step 7) reconciles a concrete
+extracted ledger instead of eyeballing two files.
+
+Usage:
+    python3 audit_translation.py <spec-file.md> <GeneratedTests.swift>
+
+It does ZERO semantic judgement — only mechanical extraction with regex, so the same
+inputs always give the same report:
+
+  1. Test-ID coverage. Every `**Test ID**: \`<id>\`` in the spec vs every `// UTS: <id>`
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply .claude/skills/uts-to-swift/scripts/audit_translation.py]: Fixed in 7e383be3 — the backslash-backtick escapes are gone; the docstring now reads `**Test ID**: <id>` marker (plain phrasing rather than nested backticks). Worth noting the same escapes exist in ably-java's `uts-to-kotlin` original — upstream candidate.

File: .claude/skills/uts-to-swift/scripts/audit_translation.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+# Swift markers ------------------------------------------------------------
+UTS_TAG_RE = re.compile(r"//\s*UTS:\s*(\S+)")
+# Swift Testing macros + the UTS infra's wait/poll helpers, both tiers. Longest
+# alternatives first so `pollUntil` isn't half-matched by `poll`.
+SWIFT_ASSERT_RE = re.compile(
+    r"(#expect|#require|"
+    r"\bawaitConnectionState\b|\bawaitChannelState\b|\bawaitState\b|"
+    r"\bpollUntil\b|\bpoll\b)"
+)
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply .claude/skills/uts-to-swift/scripts/audit_translation.py]: Fixed in 7e383be3 — good catch, and we confirmed the masking risk on real data before fixing: `RTN16f` has 2 spec asserts and 3 wait calls, so under the combined count both assertions could be deleted with the shortfall still reading negative. Assertions (`#expect`/`#require`) and infra wait/poll calls are now counted separately, mirroring `specAssertCount`/`specAwaitCount`: `assertionShortfall` stays the primary tripwire, and a new `awaitShortfall` (+ `swiftAwaitCalls/Count`, `swiftAwaitTotal`, `testsWithAwaitShortfall`) is documented as a softer secondary signal — custom continuation-bridged helpers (e.g. `awaitTime`) legitimately replace infra wait calls, so it's advisory rather than a gate. Verified: no false positives on the existing suites (ConnectionRecovery clean at 42 asserts / 20 waits split), the mutation test still trips, and SKILL.md Step 7 describes the split fields.

File: .claude/skills/uts-to-swift/scripts/audit_translation.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                continue
+            if not in_fence:
+                continue
+            stripped = line.strip()
+            if not stripped or stripped.startswith("#"):  # blank / pseudocode comment
+                continue
+            if cur is None:  # code before any heading in the block (rare)
+                cur = {"heading": "(preamble)", "lines": []}
+                sections.append(cur)
+            cur["lines"].append({"text": stripped, "kind": line_kind(stripped)})
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply .claude/skills/uts-to-swift/scripts/audit_translation.py]: Fixed in 7e383be3 — only a ```pseudo-labelled fence opens ledger extraction now; other-labelled and unlabelled fences are skipped to their closer, so the code matches its own docstring. We quantified the impact first: the corpus has 4,079 pseudo fences vs 5 json fences and 5 unlabelled blocks, and every non-pseudo fence sits in a file preamble (before the first Test ID), outside the parsed test blocks — so this is defensive hardening with provably zero output change today (146-file sweep: 0 spec-ledger diffs vs the previous revision).

File: .claude/skills/uts-to-swift/scripts/audit_translation.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    out = {}
+    for idx, (start, tag) in enumerate(tags):
+        end = tags[idx + 1][0] if idx + 1 < len(tags) else len(lines)
+        calls = []
+        for line in lines[start:end]:
+            if line.strip().startswith("//"):  # comment-only line — not executable
+                continue
+            for m in SWIFT_ASSERT_RE.finditer(line):
+                calls.append(m.group(1))
+        out[tag] = {"assertionCalls": calls, "assertionCount": len(calls)}
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply .claude/skills/uts-to-swift/scripts/audit_translation.py]: Not taking this one, deliberately. The realistic masking case — a whole commented-out `// #expect(...)` line or an annotated-omission comment — is already excluded by the comment-only-line skip, and the remaining cases (block comments around assertions, helper names inside string literals, trailing comments mentioning `#expect`) are ones the skill's own generation conventions never produce: spec comments become `//` lines and assertions are never quoted. The audit is explicitly documented as "a review aid, not a gate" — Step 7's line-by-line reconciliation is the actual completeness guarantee — so embedding a Swift lexer here buys almost nothing for the complexity it adds, and drifts further from the uts-to-kotlin original this is a minimal-diff port of. The split-counting fix from your other comment (7e383be3) also shrinks the residual surface: a stray helper name in a description string can now only pollute the advisory await count, not the assertion tripwire. Happy to revisit if a real translation ever produces a masked case.

File: .claude/skills/uts-to-swift/scripts/resolve_uts.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        help="add a mapping for this source module using NAME as the ably-cocoa "
+        "module base name, then resolve",
+    )
+    args = ap.parse_args()
+
+    # Validate via Path.parts so this works regardless of separator (Windows '\' as well as
+    # POSIX '/'); a hard-coded "/uts/<module>$" regex would reject otherwise-valid Windows paths.
+    module_dir = Path(args.module_dir)
+    raw = str(module_dir)
+    parts = module_dir.parts
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply .claude/skills/uts-to-swift/scripts/resolve_uts.py]: Fixed in 7e383be3 — `Path(args.module_dir).expanduser()`, so `~/…/uts/realtime` now resolves instead of failing with a misleading `DIR_NOT_FOUND` (the `~` component passed the `uts/<module>` structure check and only tripped on `is_dir()`). Same gap exists in ably-java's resolver — upstream candidate.

File: .claude/skills/uts-to-swift/scripts/audit_translation.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        in_fence = False
+        for line in lines[start:end]:
+            h = HEADING_RE.match(line)
+            if h and not in_fence:
+                cur = {"heading": h.group(1), "lines": []}
+                sections.append(cur)
+                continue
+            if FENCE_RE.match(line):
+                in_fence = not in_fence
+                continue
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply .claude/skills/uts-to-swift/scripts/audit_translation.py]: Fixed in 7e383be3 (same issue as CodeRabbit's comment on line 139) — extraction is now opened only by a ```pseudo-labelled fence; json/unlabelled fences are skipped to their closer. Corpus-verified output-neutral today: all 5 json fences and all 5 unlabelled blocks live in file preambles, outside test blocks (146-file sweep shows 0 ledger diffs), so this hardens against future specs placing fixtures inside test blocks.

File: .claude/skills/uts-to-swift/SKILL.md [hunk 7]
```diff
@@ -1,133 +1,240 @@
...
   marketplace: false
 ---
 
-# UTS to Swift
+Translate the UTS pseudocode test specs under the **module directory** `$ARGUMENTS` into runnable Swift
+tests in the ably-cocoa `UTS` test target (`Test/UTS`).
 
-Translate the UTS pseudocode test spec at `$ARGUMENTS` into a runnable Swift test in the `UTS` test target (`Test/UTS`).
+`$ARGUMENTS` is a UTS *module* directory — a directory sitting directly under `.../specification/uts/`,
+e.g. `/Users/sachinsh/ably-specification/specification/uts/objects`. Its name (`objects`, `realtime`,
```
Comment (maratal) [inline .claude/skills/uts-to-swift/SKILL.md]: Not sure if you want this path here.

File: .claude/skills/uts-to-swift/SKILL.md [hunk 8]
```diff
@@ -1,133 +1,240 @@
...
 
-Only continue to Step 1 once the file is confirmed to exist.
+## Step D — Choose which specs to translate
 
----
+The chosen tier's `specs` list (from Step A) is the candidate set — each entry already has its source `file`
+and derived `className`. Present it and ask whether to translate **all** of them or a **selected subset**.
+Then continue to Step E.
+
+## Step E — Translate only, or also evaluate?
```
Comment (maratal) [inline .claude/skills/uts-to-swift/SKILL.md:106]: Why agent forced to read a question in the skill file?

File: .claude/skills/uts-to-swift/SKILL.md (hunk 7 above)
Comment (sacOO7) [reply .claude/skills/uts-to-swift/SKILL.md]: It's example path, it will be specified when you don't provide a path, so it will say provide module spec path such as `/Users/sachinsh/ably-specification/specification/uts/objects`

File: .claude/skills/uts-to-swift/scripts/resolve_uts.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+             f"{raw!r} has no unit/ or integration/ sub-directory; "
+             f"not a valid UTS module.")
+
+    source_module = module_dir.name
+
+    if not MAPPING.is_file():
+        fail("MAPPING_NOT_FOUND", f"mapping file not found at {MAPPING}")
+    data = json.loads(MAPPING.read_text(encoding="utf-8"))
+    packages = data.setdefault("packages", {})
+    test_root = data.get("testRoot", "")
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply .claude/skills/uts-to-swift/scripts/resolve_uts.py]: Fixed in 455f4ca7 — a blank/missing/non-string `testRoot` now fails fast with a new structured `BAD_MAPPING` error (distinct from `MAPPING_NOT_FOUND`: the file exists but is unusable), and a trailing `/` is normalised. The validation deliberately runs *before* the `--create` branch so a corrupted mapping is never written back to disk. Verified all three variants (missing key, empty string, non-string) fail correctly and that `--create` against a broken mapping leaves the file untouched. Kotlin's resolver has the same silent-default gap — upstream candidate.

File: .claude/skills/uts-to-swift/scripts/audit_translation.py [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+    Comment-only lines are skipped when counting — a commented-out `// #expect(...)`
+    (or an annotated-omission comment) is NOT an assertion, and counting it would
+    silently mask a dropped one. (The spec-side parser skips `#` pseudocode comments
+    for the same reason.)"""
+    lines = read_lines(path)
+    tags = [(i, m.group(1)) for i, line in enumerate(lines) for m in [UTS_TAG_RE.search(line)] if m]
+    out = {}
+    for idx, (start, tag) in enumerate(tags):
+        end = tags[idx + 1][0] if idx + 1 < len(tags) else len(lines)
+        asserts, awaits = [], []
```
[the remark this answers is not in the document]
Comment (sacOO7) [reply .claude/skills/uts-to-swift/scripts/audit_translation.py:189]: Fixed in 455f4ca7, with one deliberate reshape of the remedy: duplicates are reported as `idCoverage.duplicateInSwift` and included in the exit-2 gate alongside missing/orphan IDs, rather than failing with a structured error (exit 64). A duplicated tag is a *finding* in the generated file, not an inability to run — the tool's contract is never-crash / degrade / exit 2 on gateable gaps, and this keeps duplicates consistent with that. Blocks are deliberately not merged either (merging would paper over the defect the report exists to expose). Verified with a mutation: duplicating one tag exits 2 flagging both the shared tag and the id the second method should have carried. Also confirmed against the whole spec corpus that no spec file contains duplicate full Test IDs — the same spec *point* across many test cases is common (e.g. RSP3 ×13) but tags compare full structured IDs (`<module>/<tier>/<point>/<slug>-<index>`), so legitimate overlap can't false-positive. SKILL.md's Step 7 coverage checklist gained the matching gate. Kotlin's script shares the overwrite behaviour — upstream candidate.

File: .claude/skills/uts-to-swift/SKILL.md (hunk 7 above)
Comment (sacOO7) [reply .claude/skills/uts-to-swift/SKILL.md]: Resolved in 455f4ca7 — you were right that a personal machine path didn't belong in a committed file, even as an example. Both occurrences now use `<cloned-ably-specification-repo-path>/uts/objects`, which also makes the prerequisite explicit (the usage message now links to https://github.com/ably/specification and says the path points inside a local clone). Note `uts/` sits at the spec repo root — the earlier `…/specification/uts/…` form was an artefact of the local checkout's directory name. The kotlin skill in ably-java has the same two personal paths; we've fixed them locally there too, to be pushed with the next ably-java skill sync.

File: .claude/skills/uts-to-swift/SKILL.md (hunk 8 above)
Comment (sacOO7) [reply .claude/skills/uts-to-swift/SKILL.md:106]: Deliberate, and inherited byte-for-byte from ably-java's `uts-to-kotlin` — this skill mirrors it heading-for-heading so the two SDKs' translation workflows stay diffable against each other. Step E is the one selection step whose answer the agent cannot infer on its own (whether the SDK implementation exists yet), so the heading names the exact question the agent must relay: the section's first instruction is "Ask the user which they want", and the two bullets underneath are the answer's consequences (translate-only stops after compile + review; translate-and-evaluate additionally runs Step 6). Steps C/D are imperative because the agent presents options it already computed; E is interrogative because the user owns the answer. Happy to rename it to something like "Step E — Choose the mode" if you prefer — but we'd want to do it in ably-java's skill in the same pass to keep the parity contract intact.

**Review by ttypic — APPROVED** [review]
LGTM
